### ISSUE TITLE
Issue #8: Implement Advanced Machine Learning Integration (TMLE, Super Learner)

### ISSUE THREAD

**Issue by datablogin** [datablogin/causal-inference-marketing#39]
## Summary
Implement advanced machine learning integration for causal inference, including Targeted Maximum Likelihood Estimation (TMLE), Super Learner ensemble methods, and doubly robust machine learning estimators. This enables robust causal inference with high-dimensional data and flexible model specifications.

## Review
Section 21.6 of Whatif.txt covers the TMLE. Python libraries PyTMLE and causalml have implementations of TMLE, super learner, and ensemble approaches. Packages: https://github.com/SCAI-BIO/pytmle and https://github.com/uber/causalml

## Background
Modern causal inference increasingly relies on machine learning methods to handle high-dimensional confounders and flexible model specifications. TMLE and Super Learner provide principled approaches to combine machine learning with causal inference while maintaining proper statistical properties and asymptotic guarantees.

### Core Implementation
- [ ] Create `TMLEEstimator` class for Targeted Maximum Likelihood Estimation
- [ ] Implement `SuperLearner` ensemble method for nuisance parameter estimation
- [ ] Create `DoublyRobustMLEstimator` for ML-based doubly robust estimation
- [ ] Cross-fitting (cross-validation) infrastructure for bias reduction
- [ ] Support for high-dimensional covariate spaces

### TMLE Implementation
- [ ] **One-step TMLE**: Basic targeted maximum likelihood estimation
- [ ] **Iterative TMLE**: Multi-step targeting for complex parameters
- [ ] **Longitudinal TMLE**: Extension to time-varying treatments
- [ ] **Efficient Influence Function**: Proper variance estimation
- [ ] **Cross-validation**: K-fold cross-fitting for overfitting protection

### Super Learner Framework
- [ ] **Base Learners**: Integration with scikit-learn estimators
- [ ] **Ensemble Methods**: Stacking, weighted averaging, discrete selection
- [ ] **Cross-validation**: Proper model selection and validation
- [ ] **Custom Learners**: Framework for adding domain-specific models
- [ ] **Screening**: Variable selection for high-dimensional settings

### ML Estimators
- [ ] Random Forest, Gradient Boosting, Neural Networks
- [ ] Lasso, Ridge, Elastic Net regularization
- [ ] Support Vector Machines, Gaussian Processes
- [ ] Deep learning integration (optional dependency)
- [ ] Automated hyperparameter tuning

### Enhanced Features
- [ ] Variable importance and model interpretation
- [ ] High-dimensional confounder handling (p >> n)
- [ ] Regularization path analysis
- [ ] Model stability and selection diagnostics
- [ ] Computational efficiency optimizations

### API Design
```python
from causal_inference.estimators.tmle import TMLEEstimator
from causal_inference.ml.super_learner import SuperLearner

# Define Super Learner library
sl_library = SuperLearner(
    base_learners=[
        'linear_regression',
        'random_forest', 
        'gradient_boosting',
        'neural_network',
        'lasso'
    ],
    ensemble_method='stacking',
    cv_folds=5
)

# TMLE with Super Learner
estimator = TMLEEstimator(
    outcome_learner=sl_library,
    propensity_learner=sl_library,
    cross_fitting=True,
    cv_folds=5,
    targeted_regularization=True
)

estimator.fit(treatment, outcome, covariates)
effect = estimator.estimate_ate()

# Get ML diagnostics
sl_results = estimator.get_super_learner_results()
cv_performance = estimator.get_cross_validation_performance()
variable_importance = estimator.get_variable_importance()
```

### Statistical Performance KPIs
1. **Bias Reduction**: ≥20% bias reduction vs single learner methods
2. **Efficiency Gains**: Tighter confidence intervals than non-ML methods
3. **High-Dimensional Performance**: Handle p > 0.8*n effectively
4. **Cross-Validation**: Proper CV performance without overfitting

### Computational Performance KPIs
1. **Training Speed**: Complete TMLE on 10,000 observations in <5 minutes
2. **Parallel Processing**: Efficient parallelization of cross-fitting
3. **Memory Management**: Handle high-dimensional data within memory constraints
4. **Scalability**: Sublinear scaling with ensemble size

### ML Quality KPIs
1. **Ensemble Performance**: Super Learner beats best single learner by ≥2%
2. **Variable Selection**: Effective screening in high-dimensional settings
3. **Hyperparameter Optimization**: Automated tuning improves performance
4. **Model Interpretability**: Meaningful variable importance measures

### Test Case 1: TMLE with Super Learner
```python
# Generate high-dimensional confounded data
np.random.seed(42)
n = 2000
p = 100  # High-dimensional covariates

# Generate confounders with sparse structure
true_confounders = np.random.randn(n, 10)  # Only 10 truly important
noise_confounders = np.random.randn(n, p-10)  # 90 noise variables
X = np.hstack([true_confounders, noise_confounders])

# Treatment depends on first 5 confounders
treatment_coef = np.zeros(p)
treatment_coef[:5] = [0.5, -0.3, 0.4, -0.2, 0.6]
logit_p = X @ treatment_coef
treatment_prob = 1 / (1 + np.exp(-logit_p))
treatment = np.random.binomial(1, treatment_prob)

# Outcome depends on treatment + different confounders
outcome_coef = np.zeros(p)
outcome_coef[2:8] = [0.3, -0.4, 0.5, -0.3, 0.2, 0.4]  # Overlaps with treatment model
true_ate = 2.0
outcome = X @ outcome_coef + true_ate * treatment + np.random.normal(0, 1, n)

# Prepare data
treatment_data = TreatmentData(values=treatment, treatment_type="binary")
outcome_data = OutcomeData(values=outcome, outcome_type="continuous")
covariate_data = CovariateData(values=pd.DataFrame(X), names=[f'X{i}' for i in range(p)])

# TMLE with Super Learner
estimator = TMLEEstimator(
    outcome_learner=SuperLearner([
        'linear_regression', 'lasso', 'random_forest', 'gradient_boosting'
    ]),
    propensity_learner=SuperLearner([
        'logistic_regression', 'lasso_logistic', 'random_forest', 'gradient_boosting'  
    ]),
    cross_fitting=True,
    cv_folds=5
)

estimator.fit(treatment_data, outcome_data, covariate_data)
effect = estimator.estimate_ate()

# Should recover true ATE within reasonable bounds
assert abs(effect.ate - true_ate) < 0.5  # Within 0.5 units
assert effect.ate_ci_lower < true_ate < effect.ate_ci_upper  # CI coverage
```

### Test Case 2: Super Learner Performance
```python
# Test Super Learner ensemble performance
sl = SuperLearner(
    base_learners=['linear_regression', 'lasso', 'random_forest', 'svm'],
    ensemble_method='stacking',
    cv_folds=5
)

# Fit on outcome prediction task
sl.fit(X, outcome)
predictions = sl.predict(X)

# Get individual learner performance
learner_performance = sl.get_learner_performance()
ensemble_performance = sl.get_ensemble_performance()

# Ensemble should beat best individual learner
best_individual = max(learner_performance.values())
assert ensemble_performance['mse'] <= best_individual * 0.98  # At least 2% improvement
```

### Test Case 3: Cross-Fitting Bias Reduction
```python
# Compare TMLE with and without cross-fitting
estimator_no_cf = TMLEEstimator(
    outcome_learner='random_forest',
    propensity_learner='random_forest', 
    cross_fitting=False
)

estimator_with_cf = TMLEEstimator(
    outcome_learner='random_forest',
    propensity_learner='random_forest',
    cross_fitting=True,
    cv_folds=5
)

# Fit both estimators
estimator_no_cf.fit(treatment_data, outcome_data, covariate_data)
estimator_with_cf.fit(treatment_data, outcome_data, covariate_data)

effect_no_cf = estimator_no_cf.estimate_ate()
effect_with_cf = estimator_with_cf.estimate_ate()

# Cross-fitting should reduce bias (be closer to true ATE)
bias_no_cf = abs(effect_no_cf.ate - true_ate)
bias_with_cf = abs(effect_with_cf.ate - true_ate)
assert bias_with_cf <= bias_no_cf * 1.1  # Cross-fitting should not be much worse
```

### Test Case 4: Variable Importance and Selection
```python
# Test variable importance calculation
variable_importance = estimator.get_variable_importance()

# Important variables should be ranked higher
important_vars = [f'X{i}' for i in range(8)]  # Variables that affect treatment or outcome
top_variables = variable_importance.head(10)['variable'].tolist()

# At least half of top 10 should be truly important
important_in_top = sum(1 for var in top_variables if var in important_vars)
assert important_in_top >= 5  # At least 50% correct identification
```

### Test Case 5: Computational Efficiency
```python
import time

# Test computational performance
start_time = time.time()

estimator_efficient = TMLEEstimator(
    outcome_learner=SuperLearner(['lasso', 'random_forest']),  # Smaller ensemble
    propensity_learner=SuperLearner(['lasso_logistic', 'random_forest']),
    cross_fitting=True,
    cv_folds=3,  # Fewer folds for speed
    parallel=True,
    n_jobs=4
)

estimator_efficient.fit(treatment_data, outcome_data, covariate_data)
fitting_time = time.time() - start_time

# Should complete within reasonable time
assert fitting_time < 300  # Less than 5 minutes

# Test memory usage (approximate)
import psutil
process = psutil.Process()
memory_usage = process.memory_info().rss / 1024 / 1024  # MB
assert memory_usage < 2000  # Less than 2GB
```

### Test Case 6: Longitudinal TMLE (Advanced)
```python
# Test TMLE with time-varying treatments (if implemented)
# This would be an advanced feature
# For now, placeholder for future development
pass
```

## Implementation References
- van der Laan, M. J. & Rose, S. (2011). "Targeted Learning: Causal Inference and Statistical Learning"
- van der Laan, M. J. et al. (2007). "Super Learner"
- Chernozhukov, V. et al. (2018). "Double/debiased machine learning for treatment and structural parameters"
- Kennedy, E. H. (2020). "Towards optimal doubly robust estimation of heterogeneous causal effects"
- Polley, E. C. et al. (2011). "Super learning for right-censored data"

## Definition of Done
- [ ] TMLE estimator implemented with proper targeting step
- [ ] Super Learner framework with multiple base learners
- [ ] Cross-fitting infrastructure for bias reduction
- [ ] High-dimensional data handling and variable selection
- [ ] Parallel processing for computational efficiency
- [ ] Variable importance and model interpretation tools
- [ ] Integration with existing library architecture
- [ ] Performance benchmarks meet specified KPIs

## Acceptance Criteria
1. TMLE recovers true effects within 10% on high-dimensional simulated data
2. Super Learner ensemble outperforms best individual learner by ≥2%
3. Cross-fitting provides bias reduction compared to non-cross-fitted estimates
4. Variable importance correctly identifies relevant confounders
5. Computational performance meets KPIs for high-dimensional problems
6. Memory usage remains reasonable for standard high-dimensional analyses
7. Integration maintains consistency with existing estimator interfaces

## Priority: P7 (Lowest)
While powerful for high-dimensional problems, advanced ML integration represents specialized methodology that should be implemented after all core causal inference methods are complete and the library is mature.
